docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+
+```sql
+-- DOMAIN: comet.uint64
+
+CREATE DOMAIN comet.uint64
+    AS numeric;
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_max CHECK (VALUE <= '18446744073709551615'::numeric);
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_positive CHECK (VALUE >= 0::numeric);
+```
+Additional domains for `uint32` and `uint8` will also be implemented.
+
+##### Schema migration
+
+Another point to consider is when the data structures change across CometBFT releases. There are a couple of ways to provide a solution to this
+problem.
+
+One way is to support a mechanism to migrate the old data to the new data structures. This would require additional logic for the migration process
+that would need to be run before ingesting data from an upgraded full node that contains the new data structures. But this approach might also
+cause some issues, for example, if the new data structure has a new field that there's no corresponding value in the "old" data structure, probably
+the value would need to be set to `[null]` but this can have unintended consequences.
+
+Another potential solution for this scenario is to find a way in the database that can support "versioning" of data structures. For example, let's
+assume there's a `Block` structure, let's call it `v1`. If in the future there's a need to modify this structure that is not compatible with the
+previous data structure, then the database would support a `v2` schema for `ResultBlock` and an `index` table could determine the criteria on which
+data structure should be used for inserting or querying data.
+
+There's also a possibility that the data structure could be stored in parallel with the normalized data but in one blob, e.g. in a `jsonb` field
+that contains all the information needed to return the client, this could be an optimized way to serve the data but could add to the storage
+requirements. This could improve the performance for reading data from the database. An alternative approach to this problem that can offer
+similar results would be to use a "caching" layer.
+
+
+### RPC server instance
+
+The **RPC server instance** is a node that runs the RPC API process for the data companion. This server instance provides an RPC API (v1) with
+the same JSONRPC methods in the full node endpoint. The RPC Companion service will expose the same JSONRPC methods and will accept the same request types and
+return wire compatible responses (should match the same response as the equivalent full node RPC endpoint).
+
+The **RPC server instance**, when serving a particular request, retrieves the required data from the database in order to
+fulfill the request. The data should be serialized in a way that makes it wire compatible with the CometBFT RPC endpoint.
+
+Identical requests should return idempotent responses, no side effects should cause the RPC service to return different responses.
```

Comment:
This is not true for all methods (e.g. methods that have height as param, but the param is left blank, should retrieve the data for the latest height)

Reply:
Good point.
Also, while requests can be idempotent I don't think it makes sense to talk about "idempotent responses". 
Maybe "requests are idempotent, even if some parameters are determined during the request processing itself" or something like it.

Reply:
I've just removed this section for now, it was causing more confusion than benefits. I think it's OK to leave it out.